### DISCUSSION TITLE
Speed up rippled unit tests (C++)

### DISCUSSION THREAD

**Discussion by intelliot** [XRPLBounties/Proposals#112] [Ideas]
[rippled](https://github.com/XRPLF/rippled) unit tests are very slow.

- As more unit tests are added, running all of the unit tests becomes increasingly onerous for some developers.
- Particularly slow in debug builds.
- Need to profile and pinpoint causes, and then improve those places.
- This is tech debt that anyone can pick up.

In [GitHub Actions](https://github.com/XRPLF/rippled/actions/runs/4622792000):
- Total duration **1h 2m 13s**
- test (macos-12, Ninja, Release) takes **1h 0m 28s**
- test (ubuntu-latest, Ninja, Release) takes **1h 1m 59s**

Reduce the runtime of unit tests without harming test coverage and the detection of bugs or regressions.

Potential areas to explore:

- There seems to be a delay in unit tests in `Env.cpp`, when `~AppBundle` calls `thread.join()`. Of course, the real problem is elsewhere. The actual issue is not expected to be the `std::thread::join` in `~AppBundle`, but rather the `signalStop` call that causes `ApplicationImp` to signal and waits for its children to stop before it, itself, stops.
- It may be smart to be able to select out an agreed subset of all the tests that gives good coverage, and have an easy way to invoke just that subset. Example:

```
--unittest            // runs all tests
--unittest=smokeTest  // runs smoke test subset of tests
```

- Consider how the tests are selected and marked.
- Fix https://github.com/XRPLF/rippled/issues/3432.

**Comment by jonathanlei** [comment]
Please comment in the thread below if you are interested in working on it!
